src/OrleansProviders/Streams/Memory/MemoryBatchContainer.cs

```diff
@@ -9,10 +7,10 @@
 
 namespace Orleans.Providers
 {
-    [Serializable]
     internal class MemoryBatchContainer<TSerializer> : IBatchContainer
-        where TSerializer : IMemoryMessageBodySerializer, new()
+        where TSerializer : IMemoryMessageBodySerializer
     {
+        private readonly IMemoryMessageBodySerializer serializer;
```

Comment:
to sum up a very long discussion we had in private chat, I think this extensibility point is too hacky, since it requires the *end user* to implement the hack if he wants to provide a different IMemoryMessageBodySerializer type.
IMO, we could have something that we call after MemoryBatchContainer is deserialized that augments it with the IServiceProvider (either [OnDeserialized] if it reliably works or something we explicitly call, since we control the usage of this class) and the it can use ActivatorUtilities to new up the serializer.
Alternatively, I agree that the previous approach that Jason suggested is a little more explicit on how to extend it, although I would pass IServiceProvider instead of SerializationManager, so alternative implementations can avoid using SerializationManager, which was probably a bad choice for the default serialization mechanism.